When I run this against Onivim 2 (with a resolution: `"libvim": "link:../libvimsrc"`), I get this error when commenting (using `gcc` in a Reason file):
![image](https://user-images.githubusercontent.com/13532591/77464934-2aaa2500-6dc5-11ea-88d8-af18d672f5f8.png)

If I then uncomment, I get a crash here:
```
[ERROR]    +0ms Oni2.Exception : Exception (Invalid_argument "index out of bounds"):
Raised by primitive operation at file "src/Core/BufferLine.re", line 144, characters 11-30
Called from file "src/Feature/Editor/BufferViewTokenizer.re", line 64, characters 29-75
Called from file "src/Feature/Editor/OverlaysView.re", line 75, characters 8-141
Called from file "src/Feature/Editor/EditorSurface.re", line 233, characters 4-308
Called from file "src/UI/EditorGroupView.re", line 148, characters 10-702
Called from file "src/UI/EditorLayoutView.re", line 54, characters 12-74
Called from file "list.ml", line 92, characters 20-23
Called from file "src/UI/EditorLayoutView.re", line 37, characters 2-583
Called from file "src/UI/EditorLayoutView.re", line 66, characters 17-46
Called from file "src/UI/EditorView.re", line 38, characters 17-43
Called from file "src/UI/Root.re", line 116, characters 8-28
Called from file "src/bin_editor/Oni2_editor.re", line 138, characters 19-47
Called from file "src/Core/Tick.re", line 112, characters 9-15
Called from file "src/Core/Tick.re", line 90, characters 15-32
Called from file "list.ml", line 92, characters 20-23
Called from file "list.ml", line 92, characters 32-39
Called from file "src/Core/Tick.re", line 99, characters 22-52
Called from file "src/Core/App.re", line 236, characters 4-23
Called from file "src/sdl2.re", line 648, characters 10-20
Called from file "src/bin_editor/Oni2_editor.re", line 240, characters 2-18
```

--- a/src/apitest/toggle_comment.c
+++ b/src/apitest/toggle_comment.c
@@ -88,9 +88,6 @@ MU_TEST(test_toggle_there_and_back_again_visual_multi)
 {
   vimInput("V");
   vimInput("j");
-
-  mu_check(vimCursorGetLine() == 2);
-
   vimInput("g");
   vimInput("c");
 
@@ -110,9 +107,6 @@ MU_TEST(test_toggle_there_and_back_again_visual_multi)
 
   vimInput("V");
   vimInput("j");
-
-  mu_check(vimCursorGetLine() == 2);
-
   vimInput("g");
   vimInput("c");
 
@@ -246,6 +240,54 @@ MU_TEST(test_undo_visual_multi)
   mu_check(strcmp(line3, "This is the third line of a test file") == 0);
 }
 
+MU_TEST(test_cursor_toggle_there_and_back_again)
+{
+  vimInput("g");
+  vimInput("c");
+  vimInput("c");
+
+  mu_check(vimCursorGetLine() == 1);
+  mu_check(vimCursorGetColumn() == 0);
+
+  // and back again
+
+  vimInput("g");
+  vimInput("c");
+  vimInput("c");
+
+  mu_check(vimCursorGetLine() == 1);
+  mu_check(vimCursorGetColumn() == 0);
+}
+
+MU_TEST(test_cursor_toggle_there_and_back_again_visual_multi)
+{
+  vimInput("V");
+  vimInput("j");
+
+  mu_check(vimCursorGetLine() == 2);
+  mu_check(vimCursorGetColumn() == 0);
+
+  vimInput("g");
+  vimInput("c");
+
+  mu_check(vimCursorGetLine() == 1);
+  mu_check(vimCursorGetColumn() == 0);
+
+  // and back again
+
+  vimInput("V");
+  vimInput("j");
+
+  mu_check(vimCursorGetLine() == 2);
+  mu_check(vimCursorGetColumn() == 0);
+
+  vimInput("g");
+  vimInput("c");
+
+  mu_check(vimCursorGetLine() == 1);
+  mu_check(vimCursorGetColumn() == 0);
+}
+
 MU_TEST_SUITE(test_suite)
 {
   MU_SUITE_CONFIGURE(&test_setup, &test_teardown);
@@ -259,6 +301,8 @@ MU_TEST_SUITE(test_suite)
   MU_RUN_TEST(test_set_line_comment_multi_there_and_back_again);
   MU_RUN_TEST(test_undo);
   MU_RUN_TEST(test_undo_visual_multi);
+  MU_RUN_TEST(test_cursor_toggle_there_and_back_again);
+  MU_RUN_TEST(test_cursor_toggle_there_and_back_again_visual_multi);
 }
 
 int main(int argc, char **argv)

--- a/src/normal.c
+++ b/src/normal.c
@@ -387,7 +387,6 @@ void toggle_comment(linenr_T lnum)
     mch_memmove(newp, comment, (size_t)commentlen);
     mch_memmove(newp + commentlen, line, (size_t)(linelen + 1));
     ml_replace(lnum, newp, FALSE);
-    inserted_bytes(lnum, 0, commentlen);
   }
 }
 
@@ -404,13 +403,13 @@ void toggle_comment_lines(linenr_T start, linenr_T end)
   }
 
   for (lnum = start; lnum <= end; lnum++)
-  {
     toggle_comment(lnum);
-  }
 
   // set cursoor to beginning
   curwin->w_cursor.lnum = start;
   curwin->w_cursor.col = 0;
+
+  changed_lines(start, 0, end + 1, 0);
 }
 
 /* Number of commands in nv_cmds[]. */

--- a/src/normal.c
+++ b/src/normal.c
@@ -102,6 +102,7 @@ static void nv_g_cmd(cmdarg_T *cap);
 static void nv_dot(cmdarg_T *cap);
 static void nv_redo(cmdarg_T *cap);
 static void nv_Undo(cmdarg_T *cap);
+static void nv_c(cmdarg_T *cap);
 static void nv_tilde(cmdarg_T *cap);
 static void nv_operator(cmdarg_T *cap);
 #ifdef FEAT_EVAL
@@ -278,7 +279,7 @@ static const struct nv_cmd
     {'`', nv_gomark, NV_NCH_ALW, FALSE},
     {'a', nv_edit, NV_NCH, 0},
     {'b', nv_bck_word, 0, 0},
-    {'c', nv_operator, 0, 0},
+    {'c', nv_c, 0, 0},
     {'d', nv_operator, 0, 0},
     {'e', nv_wordcmd, 0, FALSE},
     {'f', nv_csearch, NV_NCH_ALW | NV_LANG, FORWARD},
@@ -6784,6 +6785,22 @@ static void nv_Undo(cmdarg_T *cap)
     u_undoline();
     curwin->w_set_curswant = TRUE;
   }
+}
+
+/*
+ * Handle "c" command.
+ */
+static void nv_c(cmdarg_T *cap)
+{
+  /* In Visual mode and typing "gcc" triggers an operator */
+  if (cap->oap->op_type == OP_COMMENT || VIsual_active)
+  {
+    /* translate "gcc" to "gcgc" */
+    cap->cmdchar = 'g';
+    cap->nchar = 'c';
+  }
+
+  nv_operator(cap);
 }
 
 /*

--- a/src/normal.c
+++ b/src/normal.c
@@ -371,7 +371,6 @@ void toggle_comment(linenr_T lnum)
 
     mch_memmove(newp, line + commentlen, (size_t)((linelen - commentlen) + 1));
     ml_replace(lnum, newp, FALSE);
-    curwin->w_cursor.col -= commentlen;
   }
   else
   {
@@ -389,7 +388,6 @@ void toggle_comment(linenr_T lnum)
     mch_memmove(newp + commentlen, line, (size_t)(linelen + 1));
     ml_replace(lnum, newp, FALSE);
     inserted_bytes(lnum, 0, commentlen);
-    curwin->w_cursor.col += commentlen;
   }
 }
 
@@ -397,6 +395,7 @@ void toggle_comment_lines(linenr_T start, linenr_T end)
 {
   linenr_T lnum;
 
+  // if end is before start, normalize by swapping
   if (start > end)
   {
     lnum = start;
@@ -408,6 +407,10 @@ void toggle_comment_lines(linenr_T start, linenr_T end)
   {
     toggle_comment(lnum);
   }
+
+  // set cursoor to beginning
+  curwin->w_cursor.lnum = start;
+  curwin->w_cursor.col = 0;
 }
 
 /* Number of commands in nv_cmds[]. */

